### ISSUE TITLE
When I add focal loss, the time of evaluation is very slow

### ISSUE THREAD

**Issue by abc8350712** [dengdan/seglink#34]
Sorry to bother you.
I change the loss to be focal loss,it trains normally.
However,when i try to evaluate the effect of it, it's so slow and i can't understand why.
This is the code:

<code>
<pre>
    def focal_loss(self, onehot_labels, cls_preds,
                   alpha=0.25, gamma=2.0, name=None, scope=None):
        with tf.name_scope(scope, 'focal_loss', [cls_preds, onehot_labels]) as sc:
            logits = tf.convert_to_tensor(cls_preds)
            onehot_labels = tf.convert_to_tensor(onehot_labels)

            precise_logits = tf.cast(logits, tf.float32) if (
                logits.dtype == tf.float16) else logits
            onehot_labels = tf.cast(onehot_labels, precise_logits.dtype)
            predictions = tf.nn.sigmoid(precise_logits)
            predictions_pt = tf.where(tf.equal(onehot_labels, 1), predictions, 1. - predictions)
            # add small value to avoid 0
            epsilon = 1e-8
            alpha_t = tf.scalar_mul(alpha, tf.ones_like(onehot_labels, dtype=tf.float32))
            alpha_t = tf.where(tf.equal(onehot_labels, 1.0), alpha_t, 1 - alpha_t)
            losses = tf.reduce_mean(-alpha_t * tf.pow(1. - predictions_pt, gamma) * tf.log(predictions_pt + epsilon),
                                   name=name)
            return losses

    def build_loss(self, seg_labels, seg_offsets, link_labels, do_summary=True):
        batch_size = config.batch_size_per_gpu

        # note that for label values in both seg_labels and link_labels:
        #    -1 stands for negative
        #     1 stands for positive
        #     0 stands for ignored
        def get_pos_and_neg_masks(labels):
            if config.train_with_ignored:
                pos_mask = labels >= 0
                neg_mask = tf.logical_not(pos_mask)
            else:
                pos_mask = tf.equal(labels, 1)
                neg_mask = tf.equal(labels, -1)

            return pos_mask, neg_mask

        def OHNM_single_image(scores, n_pos, neg_mask):
            """Online Hard Negative Mining.
                scores: the scores of being predicted as negative cls
                n_pos: the number of positive samples
                neg_mask: mask of negative samples
                Return:
                    the mask of selected negative samples.
                    if n_pos == 0, no negative samples will be selected.
            """

            def has_pos():
                n_neg = n_pos * config.max_neg_pos_ratio
                max_neg_entries = tf.reduce_sum(tf.cast(neg_mask, tf.int32))
                n_neg = tf.minimum(n_neg, max_neg_entries)
                n_neg = tf.cast(n_neg, tf.int32)
                neg_conf = tf.boolean_mask(scores, neg_mask)
                vals, _ = tf.nn.top_k(-neg_conf, k=n_neg)
                threshold = vals[-1]  # a negtive value
                selected_neg_mask = tf.logical_and(neg_mask, scores <= -threshold)
                return tf.cast(selected_neg_mask, tf.float32)

            def no_pos():
                return tf.zeros_like(neg_mask, tf.float32)

            return tf.cond(n_pos > 0, has_pos, no_pos)

        def OHNM_batch(neg_conf, pos_mask, neg_mask):
            selected_neg_mask = []
            for image_idx in xrange(batch_size):
                image_neg_conf = neg_conf[image_idx, :]
                image_neg_mask = neg_mask[image_idx, :]
                image_pos_mask = pos_mask[image_idx, :]
                n_pos = tf.reduce_sum(tf.cast(image_pos_mask, tf.int32))
                selected_neg_mask.append(OHNM_single_image(image_neg_conf, n_pos, image_neg_mask))

            selected_neg_mask = tf.stack(selected_neg_mask)
            selected_mask = tf.cast(pos_mask, tf.float32) + selected_neg_mask
            return selected_mask

        # OHNM on segments
        seg_neg_scores = self.seg_scores[:, :, 0]
        seg_pos_mask, seg_neg_mask = get_pos_and_neg_masks(seg_labels)
        seg_selected_mask = OHNM_batch(seg_neg_scores, seg_pos_mask, seg_neg_mask)
        n_seg_pos = tf.reduce_sum(tf.cast(seg_pos_mask, tf.float32))

        with tf.name_scope('seg_cls_loss'):
            def has_pos():
                #seg_cls_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
                #    logits=self.seg_score_logits,
                #    labels=tf.cast(seg_pos_mask, dtype=tf.int32))
                #return tf.reduce_sum(seg_cls_loss * seg_selected_mask) / n_seg_pos
                seg_cls_loss = self.focal_loss(tf.one_hot(seg_labels, 2), self.seg_score_logits)
                return seg_cls_loss
            def no_pos():
                return tf.constant(.0);

            seg_cls_loss = tf.cond(n_seg_pos > 0, has_pos, no_pos)
            tf.add_to_collection(tf.GraphKeys.LOSSES, seg_cls_loss)

        def smooth_l1_loss(pred, target, weights):
            diff = pred - target
            abs_diff = tf.abs(diff)
            abs_diff_lt_1 = tf.less(abs_diff, 1)
            if len(target.shape) != len(weights.shape):
                loss = tf.reduce_sum(tf.where(abs_diff_lt_1, 0.5 * tf.square(abs_diff), abs_diff - 0.5), axis=2)
                return tf.reduce_sum(loss * tf.cast(weights, tf.float32))
            else:
                loss = tf.where(abs_diff_lt_1, 0.5 * tf.square(abs_diff), abs_diff - 0.5)
                return tf.reduce_sum(loss * tf.cast(weights, tf.float32))

        with tf.name_scope('seg_loc_loss'):
            def has_pos():
                seg_loc_loss = smooth_l1_loss(self.seg_offsets, seg_offsets,
                                              seg_pos_mask) * config.seg_loc_loss_weight / n_seg_pos
                names = ['loc_cx_loss', 'loc_cy_loss', 'loc_w_loss', 'loc_h_loss', 'loc_theta_loss']
                sub_loc_losses = []
                from tensorflow.python.ops import control_flow_ops
                for idx, name in enumerate(names):
                    name_loss = smooth_l1_loss(self.seg_offsets[:, :, idx], seg_offsets[:, :, idx],
                                               seg_pos_mask) * config.seg_loc_loss_weight / n_seg_pos
                    name_loss = tf.identity(name_loss, name=name)
                    if do_summary:
                        tf.summary.scalar(name, name_loss)
                    sub_loc_losses.append(name_loss)
                seg_loc_loss = control_flow_ops.with_dependencies(sub_loc_losses, seg_loc_loss)
                return seg_loc_loss

            def no_pos():
                return tf.constant(.0);

            seg_loc_loss = tf.cond(n_seg_pos > 0, has_pos, no_pos)
            tf.add_to_collection(tf.GraphKeys.LOSSES, seg_loc_loss)

        link_neg_scores = self.link_scores[:, :, 0]
        link_pos_mask, link_neg_mask = get_pos_and_neg_masks(link_labels)
        link_selected_mask = OHNM_batch(link_neg_scores, link_pos_mask, link_neg_mask)
        n_link_pos = tf.reduce_sum(tf.cast(link_pos_mask, dtype=tf.float32))
        with tf.name_scope('link_cls_loss'):
            def has_pos():
                #link_cls_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
                #    logits=self.link_score_logits,
                #    labels=tf.cast(link_pos_mask, tf.int32))
                #return tf.reduce_sum(link_cls_loss * link_selected_mask) / n_link_pos
                link_cls_loss = self.focal_loss(tf.one_hot(link_labels, 2), self.link_score_logits)
                return link_cls_loss
            def no_pos():
                return tf.constant(.0);

            link_cls_loss = tf.cond(n_link_pos > 0, has_pos, no_pos) * config.link_cls_loss_weight
            tf.add_to_collection(tf.GraphKeys.LOSSES, link_cls_loss)

        if do_summary:
            tf.summary.scalar('seg_cls_loss', seg_cls_loss)
            tf.summary.scalar('seg_loc_loss', seg_loc_loss)
            tf.summary.scalar('link_cls_loss', link_cls_loss)
</pre>
</code>

Thanks in advance.

**Comment by dengdan** [comment]
If your model had not been well trained, the post-processing steps could be very time consuming because there are some loops on positive predictions.  And therefore, It can be speculated that you focal loss model was poorly trained. 
Firstly, I see that you are using sigmoid as the probability output function, have you modified the net structure as well, since I used softmax instead.
Secondly, the focal loss has alpha, gamma as its hyper parameters, have you tried and find a good combination for this task?

**Comment by abc8350712** [comment]
It's awesome of you.I never thought that poor  trained model can lead to this problem..
Thank you very much!

**Comment by abc8350712** [comment]
It's awesome of you.I never thought that poor trained model can lead to this problem..
Thank you very much!
